module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1987 @@
+{
+	/*
+	 * The ZIL MAC is embedded in the block it protects, which will
+	 * not have been byteswapped by the time this function has been called.
+	 * As a result, we don't need to worry about byteswapping the MAC.
+	 */
+	const zil_chain_t *zilc = data;
+
+	bcopy(&zilc->zc_eck.zec_cksum.zc_word[2], mac, sizeof (uint64_t));
+	bcopy(&zilc->zc_eck.zec_cksum.zc_word[3], mac + sizeof (uint64_t),
+	    sizeof (uint64_t));
+}
+
+/*
+ * This routine takes a block of dnodes (src_abd) and copies only the bonus
+ * buffers to the same offsets in the dst buffer. datalen should be the size
+ * of both the src_abd and the dst buffer (not just the length of the bonus
+ * buffers).
+ */
+void
+zio_crypt_copy_dnode_bonus(abd_t *src_abd, uint8_t *dst, uint_t datalen)
+{
+	uint_t i, max_dnp = datalen >> DNODE_SHIFT;
+	uint8_t *src;
+	dnode_phys_t *dnp, *sdnp, *ddnp;
+
+	src = abd_borrow_buf_copy(src_abd, datalen);
+
+	sdnp = (dnode_phys_t *)src;
+	ddnp = (dnode_phys_t *)dst;
+
+	for (i = 0; i < max_dnp; i += sdnp[i].dn_extra_slots + 1) {
+		dnp = &sdnp[i];
+		if (dnp->dn_type != DMU_OT_NONE &&
+		    DMU_OT_IS_ENCRYPTED(dnp->dn_bonustype) &&
+		    dnp->dn_bonuslen != 0) {
+			bcopy(DN_BONUS(dnp), DN_BONUS(&ddnp[i]),
+			    DN_MAX_BONUS_LEN(dnp));
+		}
+	}
+
+	abd_return_buf(src_abd, src, datalen);
+}
+
+static int
+zio_crypt_do_dnode_hmac_updates(crypto_context_t ctx, boolean_t byteswap,
+    dnode_phys_t *dnp)
+{
+	int ret, i;
+	dnode_phys_t *adnp;
+	blkptr_t *curr_bp, *bp;
+	blkptr_t tmpbp;
+	boolean_t need_bswap = (byteswap ^ !ZFS_HOST_BYTEORDER);
+	uint64_t blkprop;
+	crypto_data_t cd;
+	uint8_t tmp_dncore[offsetof(dnode_phys_t, dn_blkptr)];
+	uint8_t mac[ZIO_DATA_MAC_LEN];
+
+	cd.cd_format = CRYPTO_DATA_RAW;
+	cd.cd_offset = 0;
+
+	/* authenticate the core dnode (masking out non-portable bits) */
+	bcopy(dnp, tmp_dncore, sizeof (tmp_dncore));
+	adnp = (dnode_phys_t *)tmp_dncore;
+	if (need_bswap) {
+		adnp->dn_datablkszsec = BSWAP_16(adnp->dn_datablkszsec);
+		adnp->dn_bonuslen = BSWAP_16(adnp->dn_bonuslen);
+		adnp->dn_maxblkid = BSWAP_64(adnp->dn_maxblkid);
+		adnp->dn_used = BSWAP_64(adnp->dn_used);
+	}
+	adnp->dn_flags &= DNODE_CRYPT_PORTABLE_FLAGS_MASK;
+	adnp->dn_used = 0;
+
+	cd.cd_length = sizeof (tmp_dncore);
+	cd.cd_raw.iov_base = (char *)adnp;
+	cd.cd_raw.iov_len = cd.cd_length;
+
+	ret = crypto_mac_update(ctx, &cd, NULL);
+	if (ret != CRYPTO_SUCCESS) {
+		ret = SET_ERROR(EIO);
+		goto error;
+	}
+
+	for (i = 0; i < dnp->dn_nblkptr + 1; i++) {
+		if (i < dnp->dn_nblkptr) {
+			curr_bp = &dnp->dn_blkptr[i];
+		} else if (dnp->dn_flags & DNODE_FLAG_SPILL_BLKPTR) {
+			curr_bp = DN_SPILL_BLKPTR(dnp);
+		} else {
+			break;
+		}
+
+		if (byteswap) {
+			tmpbp = *curr_bp;
+			byteswap_uint64_array(&tmpbp, sizeof (blkptr_t));
+			bp = &tmpbp;
+		} else {
+			bp = curr_bp;
+		}
+
+		blkprop = bp->blk_prop;
+		BF64_SET(blkprop, 62, 1, 0);
+		BF64_SET(blkprop, 40, 8, 0);
+		BF64_SET(blkprop, 16, 16, 0);
```

Comment:
How about a function that takes the BP and returns the props word with the nonportable bits zeroed out.  If you're worried about copying the whole BP, you could copy just the props word to a local bp, and use the BP_SET_* macros on it. (But I would argue that copying the whole BP is probably fine, unless you have data to the contrary.)